File: alignment/parse_qc.pl
#!/usr/bin/perl
use strict;
use warnings;

if(@ARGV != 2){
    print STDERR "Usage: $0 {output flag 0 - table, 1 - curl/jason command/output 2 - both}{sample prefix}\n";
    exit(1);
}
my $flag=$ARGV[0];
my $prefix=$ARGV[1];

my %files = archiveLookup();
my $tbl.="BionimbusID\tDate\tMachine\tRun\tBarCode\tLane\tread_length\ttotal_reads\tpost_align_reads\tfraction_aligned\tpost_rmdup_reads\tfraction_rmduped\ttarget_size\taligned_bp_ot\tfraction_aligned_bp_ot\tfraction_sequenced_bp_ot\taverage_ot_cov\tcoverage1\tcoverage2\tcoverage8\tfraction_coverage1\tfraction_coverage2\tfraction_coverage8\taligned_target_enrichment\tsequenced_target_enrichment\tmedian_insert_size\tmedian_absolute_deviation\tmean_insert_size\tinsert_standard_devation\n";

my %destinations;
my %runs;
my $GENOME_SIZE = 3137161264;
my $hostname='localhost';
my $port=9200;

my @samples = sort keys %files;
foreach my $sample (@samples) {
    print STDERR "$sample\n";
    my @set = sort keys %{$files{$sample}};
    
}
my %data = directory_search(".");
output_stats(\%data,\$flag);
if($flag == 0 || $flag == 2){
    open(TBL, ">","$prefix.qc_stats.txt");
    print TBL $tbl;
    close TBL;
}
sub output_stats
{
    #date variables
    my @months = qw( Jan Feb Mar Apr May Jun Jul Aug Sep Oct Nov Dec );
    my @days = qw(Sun Mon Tue Wed Thu Fri Sat Sun);
    (my $sec,my $min,my $hour,my $mday,my $mon,my $year,my $wday,my $yday,my $isdst)=gmtime();
    my $cur_date="$days[$wday], $mday $months[$mon] ".($year+1900)." $hour:$min:$sec GMT";
    
    my $data_ref = $_[0];
    my $f_ref=$_[1];
    my $f=$$f_ref;
    my %data = %$data_ref;
    my @samples = sort (keys %data);
    foreach my $sample (@samples) {
	my @annot = split (/\_/, $sample);
	my ($db_id,$date,$machine,$run,$barcode,$lane);
	$db_id = $annot[0];
	if($data{$sample}{SRC} eq 'HGAC'){
	    $date = $annot[1];
	    $machine = $annot[2];
	    $run = $annot[3];
	    $barcode = $annot[4];
	    $lane = $annot[5];
	}
	else{
	    ($date,$machine,$run,$barcode)=("NA","NA","NA","NA");
	    $lane = $annot[-1];
	}
	unless ($data{$sample}{LEN}) {
	    $data{$sample}{LEN} = 100;
	}
	unless ($data{$sample}{READS}) {
	    $data{$sample}{READS}="NA";
	}
	if ($data{$sample}{RAW}{MAP}) {
	    # add DB ID, Run date, machine SN, run #, flow cell barcode, lane
	    $tbl.="$db_id\t$date\t$machine\t$run\t$barcode\t$lane\t";
	    
	    # aggregate arguments and structure for curl/JSON command output
	    my $json.="{\n\t\"Bionimbus_id\": \"".$db_id."\",\n\t\"Run\":";
	    $json.=" \"".$run."\",\n\t\"Date\": \"".$date."\",\n\t\"Machine\": \"".$machine."\",\n\t\"BarCode\": \"".$barcode."\",\n\t\"Lane\": \"".$lane."\",\n\t\"read_length\": \"".$data{$sample}{LEN}."\",\n\t\"total_reads\": \"".$data{$sample}{RAW}{TOTAL}."\",\n";
	    # get Total reads, Aligned reads, rmdup aligned reads, Fraction duplicate reads
	    my $ali_frac = sprintf ("%.4f", $data{$sample}{RAW}{MAP}/$data{$sample}{RAW}{TOTAL});
	    my $dup_frac = sprintf ("%.4f", 1 - ($data{$sample}{RMDUP}{MAP} / $data{$sample}{RAW}{TOTAL}));
	    
	    # get # on target bases, Read length, Fraction on target bases, On target Enrichment
	    my $total_bp = $data{$sample}{RAW}{TOTAL} * $data{$sample}{LEN};
	    my $aligned_rmdup_bp = $data{$sample}{RMDUP}{MAP} * $data{$sample}{LEN};
	    my $on_target_reads = $data{$sample}{COV}{TOTAL} / $data{$sample}{LEN};
	    my $on_target_bp = $data{$sample}{COV}{TOTAL};
	    my $frac_aligned_bp_ot = sprintf ("%.4f", $on_target_bp / $aligned_rmdup_bp);
	    my $frac_sequenced_bp_ot = sprintf ("%.4f", $on_target_bp / $total_bp);
	    # add alignment stats to JSON statement
	    # median_insert_size\tmedian_absolute_deviation\tmean_insert_size\tinsert_standard_devation
	    #add date to json object
# Wed, 04 Mar 2015 01:41:44 GMT                                                                                                                                               
	    my $cur_date=$days[$wday].", $mday ".$months[$mon]." ".($year+1900)." ".$hour.":".$min.":".$sec." GMT";
	    $json.="\t\"post_align_reads\": \"".$data{$sample}{RAW}{MAP}."\",\n\t\"fraction_aligned\": \"".$ali_frac."\",\n\t\"post_rmdup_reads\": \"".$data{$sample}{RMDUP}{MAP}."\",\n\t\"fraction_rmduped\": \"".$dup_frac."\",\n\t\"target_size\": \"".$data{$sample}{COV}{TARGET}."\",\n\t\"aligned_bp_ot\": \"".$data{$sample}{COV}{TOTAL}."\",\n\t\"frac_aligned_bp_ot\": \"".$frac_aligned_bp_ot."\",\n\t\"fraction_sequenced_bp_ot\": \"".$frac_sequenced_bp_ot."\",\n\t\"median_insert_size\": \"".$data{$sample}{MI}."\",\n\t\"median_absolute_deviation\": \"".$data{$sample}{MA}."\",\n\t\"mean_insert_size\": \"".sprintf("%.2f",$data{$sample}{XI})."\",\n\t\"insert_standard_deviation\": \"".sprintf("%.2f",$data{$sample}{SI})."\",\n\t\"date_aligned\": \"".$cur_date."\",\n";
	    
	    my $average_ot_coverage = sprintf ("%.2f", $on_target_bp / $data{$sample}{COV}{TARGET});
	    my $on_target_frac = sprintf ("%.4f", $on_target_reads / $data{$sample}{RMDUP}{MAP} );
	    
	    # calculate enrichment
	    
	    my $frac_sizes = sprintf ("%.4f", $GENOME_SIZE / $data{$sample}{COV}{TARGET});
	    my $frac_aligned_reads = sprintf ("%.4f", $aligned_rmdup_bp / $on_target_bp);
	    my $aligned_target_enrichment = sprintf ("%.2f", $frac_sizes / $frac_aligned_reads);
	    
	    my $frac_sequenced_reads = sprintf ("%.4f", $total_bp / $on_target_bp);
	    my $sequenced_target_enrichment = sprintf ("%.2f", $frac_sizes / $frac_sequenced_reads);
	    
	    #HEADER: read_length total_reads post_align_reads fraction_aligned post_rmdup_reads fraction_rmduped target_size aligned_bp_ot fraction_aligned_bp_ot fraction_sequenced_bp_ot average_ot_cov coverage1 coverage2 coverage8 fraction_coverage1 fraction_coverage2 fraction_coverage8 aligned_target_enrichment sequenced_target_enrichment\n";
	    $tbl.="$data{$sample}{LEN}\t";			# read_length
	    $tbl.="$data{$sample}{RAW}{TOTAL}\t";	# total_reads
	    $tbl.="$data{$sample}{RAW}{MAP}\t";		# post_align_reads
	    $tbl.="$ali_frac\t";					# fraction_aligned
	    $tbl.="$data{$sample}{RMDUP}{MAP}\t";	# post_rmdup_reads
	    $tbl.="$dup_frac\t";					# fraction_rmduped
	    $tbl.="$data{$sample}{COV}{TARGET}\t";	# target_size
	    $tbl.="$data{$sample}{COV}{TOTAL}\t";	# aligned_bp_ot
	    $tbl.="$frac_aligned_bp_ot\t";			# fraction_aligned_bp_ot
	    $tbl.="$frac_sequenced_bp_ot\t";		# fraction_sequenced_bp_ot
	    $tbl.="$average_ot_coverage\t";			# average_ot_cov
				# add coverage stats to json
	    $json.="\t\"average_ot_coverage\": \"".$average_ot_coverage."\",\n\t";
	    for('1','2','8') {            			# cov 1x 2x 8x
		$tbl.= $data{$sample}{COV}{"RS$_"} . "\t";
		my $key="RS$_";
		$json.="\"coverage$_\": \"".$data{$sample}{COV}{$key}."\",\n\t";
	    }
	    for('1','2','8') {						# frac cov 1x 2x 8x
		my $fc = sprintf ("%.4f", $data{$sample}{COV}{"RS$_"} / $data{$sample}{COV}{TARGET});
		$tbl.= $fc."\t";
		$json.="\"fraction_coverage$_\": \"".$fc."\",\n\t";
	    }
	    $tbl.= "$aligned_target_enrichment\t";
	    $tbl.= "$sequenced_target_enrichment\t";
	    $tbl.="$data{$sample}{MI}\t$data{$sample}{MA}\t".sprintf("%.2f",$data{$sample}{XI})."\t".sprintf("%.2f",$data{$sample}{SI})."\n";
	    $json.="\"aligned_target_enrichment\": \"".$aligned_target_enrichment."\",\n\t\"sequenced_target_enrichment\": \"".$sequenced_target_enrichment."\"\n}";
	    #	    system($json);
	    if($f == 1 || $f == 2){
		open(CJ, ">>", "$prefix.qc_stats.json");
		print CJ $json;
		close CJ;
	    }
	}
    }
}
sub directory_search
{
    my %data;
    my @dir = `ls $_[0]`;
    foreach my $file (@dir) {
	chomp $file;
	if ($file =~ /(.+[1-8])\.rmdup.srt.bam.flagstats$/ || $file =~ /(.+[1-8]).rmdup.accepted_hits.bam.flagstats$/) {
	    my $sample = $1;
	    %{$data{$sample}{RMDUP}} = parseFS("$_[0]/$file");
	    my @name = split (/_/, $sample);
	    $data{$name[0]}{RUNS}{$_[0]}{$name[-1]}++;
	    
	}
	elsif ($file =~ /(.+[1-8])\.srt\.bam\.flagstats$/ || $file =~ /(.+[1-8]).accepted_hits.bam.flagstats$/) {
	    my $sample = $1;
	    %{$data{$sample}{RAW}} = parseFS("$_[0]/$file");
	}
	elsif ($file =~ /(.+[1-8]).(genome|exome|capture)\.hist$/ || $file =~ /(.+[1-8]).rnaSeq.covWhist$/) {
	    my $sample = $1;
	    ($data{$sample}{COV}{TOTAL}, $data{$sample}{COV}{RS1}, $data{$sample}{COV}{RS2}, $data{$sample}{COV}{RS8}, $data{$sample}{COV}{TARGET}) = parseCoverage("$_[0]/$file");
	}
	elsif ($file =~ /(.+[1-8])_[12]_sequence.txt.gz$/ || $file =~ /(.+[1-8])_sequence.txt.gz$/ || $file=~ /(^\S+)_\D*\d\.f\w*q\.gz$/) {
	    my $sample = $1;
	    my $temp = `gzip -dc $_[0]/$file | head -n 2 | tail -n 1`;
	    chomp $temp;
	    my $length = length($temp);
	    $data{$sample}{LEN} = $length;
	    if($file=~/_sequence.txt.gz$/){
		$data{$sample}{SRC}='HGAC';
	    }
	    else{
		$data{$sample}{SRC}='other';
	    }
	}
	elsif ($file=~/(.+[1-8])\.insert_metrics.hist$/){
	    my $sample=$1;
	    ($data{$sample}{MI},$data{$sample}{MA},$data{$sample}{XI},$data{$sample}{SI})=parseIns($file);
	}
	
    }
    return(%data);
}
sub parseFS
{
    my %data;
    my ($fs)  = @_;
    open IN, $fs or die "Can't open $fs!\n";
    my $line = <IN>;
    chomp $line;
    if($line =~ /^(\d+)\s/) {
	$data{TOTAL} = $1;
    }
    $line = <IN>;
    $line = <IN>;
    chomp $line;
    if($line =~ /^(\d+).*\((\d+\.\d+)/) {
	$data{MAP} = $1;
	$data{PCT} = $2;
    }
    return(%data);
}
sub parseCoverage
{
    my $total_bases = 0;
    my $rs1_bases = my $rs2_bases = my $rs8_bases = 0;
    my $target_size = 0;
    my ($fs)  = @_;
    open IN, $fs or die "Can't open $fs!\n";
    my $target_size_flag = 0;
    while (<IN>) {
	chomp $_;
	my @temp = split (/\s+/, $_);
	if(!$target_size_flag){
	    $target_size = $temp[3];
	    $target_size_flag = 1;
	}
	$total_bases = $total_bases + $temp[1]*$temp[2];
	$target_size = $temp[3];
	if ($temp[1] >= 1) {
	    $rs1_bases = $rs1_bases + $temp[2];
	}
	if ($temp[1] >= 2) {
	    $rs2_bases = $rs2_bases + $temp[2];
	}
	if ($temp[1] >= 8) {
	    $rs8_bases = $rs8_bases + $temp[2];
	}
    }
    return($total_bases, $rs1_bases, $rs2_bases, $rs8_bases, $target_size);
}

sub parseIns
{
    my ($f)=@_;
    open(INS, $f) or die "Can't open insert histogram file $f\n";
    while(my $line=<INS>){
	chomp $line;
	if($line=~/^MEDIAN/){
	    my $stats=<INS>;
	    close INS;
	    my @stats= split /\t/,$stats;
	    return($stats[0],$stats[1],$stats[4],$stats[5]);
	}
    }
}
sub annotIn
{
    my %annotation;
    open IN, $_[0] or die "Can't open $_[0]!\n";
    while (<IN>) {
	chomp $_;
	my @temp = split (/\t/, $_);
	my $db_id = $temp[0];
	$annotation{$db_id}{EXP_TYPE} = $temp[3];
	$annotation{$db_id}{SAMPLE_NAME} = $temp[1];
	$annotation{$db_id}{USER} = $temp[6];
	$annotation{$db_id}{PROJECT} = $temp[2];
    }
    return(%annotation);
}
sub archiveLookup
{
    my %files;
    my @dir = `ls .`;
    foreach my $file (@dir) {
	chomp $file;
	# add flexibility for fastq naming conventions
	if ($file =~ /(.+)_sequence.txt(.gz)*/) {
	    my @temp = split (/_/, $file);
	    my $db_id = $temp[0];
	    my $date = $temp[1];
	    my $mid = $temp[2];
	    my ($year, $month, $day) = (0,0,0);
	    if ($date =~ /(\d\d)(\d\d)(\d\d)/) {
		$year = $1;
		$month = $2;
		$day = $3;
	    }
	    $files{$db_id}{$file}++;
	}
	if($file =~/(^\S+)_\D*\d\.f\w*q\.gz$/){
	    my @temp = split (/_/, $file);
	    my $db_id = $temp[0];
	    my $date = $temp[1];
	    my $mid = $temp[2];
	    my ($year, $month, $day) = (0,0,0);
	    if ($date =~ /(\d\d)(\d\d)(\d\d)/) {
		$year = $1;
		$month = $2;
		$day = $3;
	    }
	    $files{$db_id}{$file}++;
	    
	}
    }
    return(%files);
}

File: alignment/parse_qc.py
#!/usr/bin/env python

import json
import os
import re
import subprocess
import sys
import time
sys.path.append('/home/ubuntu/TOOLS/Scripts/')
from utility.date_time import date_time
from utility.log import log


def parse_config(config_file):
    config_data = json.loads(open(config_file, 'r').read())
    return config_data['params']['cc_vsn'], int(config_data['params']['genome_size']), config_data['params']['ranges']


def print_header_cc(tbl, ccvsn, ranges):
    tbl.write(
        "BionimbusID\tDate\tMachine\tRun\tBarCode\tLane\tread_length\ttotal_reads\tpost_align_reads\tpercent_aligned" +
        "\tpost_rmdup_reads\tfraction_rmduped\ttarget_size\taligned_bp_ot\tfraction_aligned_bp_ot\t" +
        "fraction_sequenced_bp_ot\tmedian_insert_size\tmedian_absolute_deviation\tmean_insert_size\t" +
        "insert_standard_deviation\tdate_aligned\t" + ccvsn + "_t1_average_x_coverage\t" + ccvsn +
        "_t1_%covered_at_average\t")
    for r in ranges:
        r *= 100
        r = str(int(r))
        tbl.write(ccvsn + "_t1_" + r + "%\t")
    tbl.write(
        ccvsn + "_t1_zero_cov_bp\t" + ccvsn + "_t1_%zero_cov\t" + ccvsn + "_t2_average_x_coverage\t" + ccvsn +
        "_t2_%covered_at_average\t")
    for r in ranges:
        r *= 100
        r = str(int(r))
        tbl.write(ccvsn + "_t2_" + r + "%\t")
    tbl.write(ccvsn + "_t2_zero_cov_bp\t" + ccvsn + "_t2_%zero_cov\n")


def parseFS(FS):
    fh = open(FS, 'r')
    line = next(fh)
    line = line.rstrip('\n')
    rd_ct = re.search('^(\d+)\s', line)
    line = next(fh)
    line = next(fh)
    line = line.rstrip('\n')
    mapped = re.search('^(\d+).*\((\d+\.\d+)', line)
    fh.close()
    return rd_ct.group(1), mapped.group(1), mapped.group(2)


def parseINS(INS):
    fh = open(INS, 'r')
    for i in xrange(0, 7, 1):
        skip = next(fh)
    line = next(fh)
    line = line.rstrip('\n')
    stats = line.split('\t')
    fh.close()
    return stats[0], stats[1], stats[4], stats[5]


def parseCoverage(CF, ranges):
    fh = open(CF, 'r')
    # if run twice will end up being tier-specific
    cvg = {}
    zero_cov = next(fh)
    zero_cov.rstrip('\n')
    zero_attr = zero_cov.split('\t')
    temp = [zero_attr]
    target = int(zero_attr[3])
    zero_bp = int(zero_attr[2])
    zero_ratio = float(zero_attr[4])
    run_ratio = 1 - float(zero_attr[4])
    run_total = 0
    aln_bp = 0
    for ratio in ranges:
        cvg[ratio] = {}
        cvg[ratio]['xcov'] = 0
        cvg[ratio]['bases'] = 0
    for line in fh:
        line = line.rstrip('\n')
        attr = line.split('\t')
        temp.append(attr)
        aln_bp += (int(attr[1]) * int(attr[2]))
        run_ratio -= float(attr[4])
        run_total += int(attr[2])
        for ratio in ranges:
            if run_ratio <= ratio and cvg[ratio]['bases'] == 0:
                cvg[ratio]['bases'] = run_total
                cvg[ratio]['xcov'] = int(attr[1])
    fh.close()
    avg_cov = aln_bp / run_total
    avg_ratio = 0.0
    for i in xrange(0, len(temp), 1):
        avg_ratio += float(temp[i][4])
        # pdb.set_trace()
        if int(temp[i][1]) >= avg_cov:
            avg_ratio = 1 - avg_ratio
            break
    return target, aln_bp, run_total, cvg, zero_bp, zero_ratio, avg_cov, avg_ratio


def parse_qc(config_file, sample, cflag):
    log_dir = './'
    if os.path.isdir('LOGS'):
        log_dir = 'LOGS/'
    loc = log_dir + sample + '.qc_stats.log'
    # set up input files and variables - .hist, .flagstats, .metrics, .qs
    insert = sample + '.insert_metrics.hist'
    rawFlag = sample + '.srt.bam.flagstats'
    if os.path.isfile(rawFlag) == False:
        rawFlag = sample + '.bam.flagstats'

    rmdupFlag = sample + '.rmdup.srt.bam.flagstats'
    qs = sample + '_1.qs'
    tbl = open(sample + '.qc_stats.txt', 'w')
    json_out = open(sample + '.qc_stats.json', 'w')
    # get read length using line count of quality stats file
    fq_rd_len = subprocess.check_output('wc -l ' + qs, shell=True)
    wc_res = fq_rd_len.split()
    rd_len = int(wc_res[0]) - 1
    date_aligned = time.strftime("%c")
    (ccvsn, genome_size, range_list) = parse_config(config_file)
    ranges = range_list.split(',')
    ranges = map(float, ranges)
    # list containing read group info
    RG = sample.split('_')
    log(loc, date_time() + 'Parsing flagstats file ' + rawFlag + '\n')
    (tot_rds, post_aligned_reads, frac_aligned) = parseFS(rawFlag)
    log(loc, date_time() + 'Parsing flagstats file ' + rmdupFlag + '\n')
    (rmduped_reads, rmduped_aligned, dummy) = parseFS(rmdupFlag)
    frac_rmduped_reads = 1 - (float(rmduped_aligned) / float(post_aligned_reads))
    # estimate total base pairs covered
    all_bp_est = int(rd_len) * int(rmduped_reads)
    (target_size, aligned_bp_ot, num_bp, fraction_aligned_bp_ot, fraction_sequenced_ot) = (0, 0, 0.0, 0, 0.0)
    log(loc, date_time() + 'Parsing read insert size file ' + insert + '\n')
    (median_insert_size, median_absolute_deviation, mean_insert_size, insert_standard_deviation) = parseINS(insert)
    # output changes based on capture or whole genome
    if cflag == 'n':
        log(loc, date_time() + 'Custom capture flag given, computing stats for tier 1 and tier 2\n')
        bed_hist1 = sample + '.capture_t1.hist'
        bed_hist2 = sample + '.capture_t2.hist'
        print_header_cc(tbl, ccvsn, ranges)
        tbl.write('\t'.join(RG) + '\t' + str(rd_len))
        # cov is a dictionary with the per -ratio stats
        log(loc, date_time() + 'Parsing bedtools coverage file ' + bed_hist1 + '\n')
        (
            t1_ts, t1_aln_bp_ot, t1_num_bp, t1_cov, t1_zero_cov, t1_pct_zero_cov, t1_avg_cov,
            t1_avg_ratio) = parseCoverage(
            bed_hist1, ranges)
        target_size += t1_ts
        aligned_bp_ot += t1_aln_bp_ot
        num_bp += t1_num_bp
        log(loc, date_time() + 'Parsing bedtools coverage file ' + bed_hist2 + '\n')
        (
            t2_ts, t2_aln_bp_ot, t2_num_bp, t2_cov, t2_zero_cov, t2_pct_zero_cov, t2_avg_cov,
            t2_avg_ratio) = parseCoverage(
            bed_hist2, ranges)
        target_size += t2_ts
        aligned_bp_ot += t2_aln_bp_ot
        num_bp += t2_num_bp
        fraction_aligned_bp_ot = float(num_bp) / float(target_size)
        fraction_sequenced_ot = float(aligned_bp_ot) / float(all_bp_est)
        # store items to print as a list so that they can be cast as a string in a smoother statement
        to_print = (
            tot_rds, post_aligned_reads, frac_aligned, rmduped_reads, frac_rmduped_reads, target_size, aligned_bp_ot,
            fraction_aligned_bp_ot, fraction_sequenced_ot, median_insert_size, median_absolute_deviation,
            mean_insert_size,
            insert_standard_deviation, date_aligned, t1_avg_cov, t1_avg_ratio)
        tbl.write('\t' + '\t'.join(str(e) for e in to_print))
        for ratio in ranges:
            tbl.write('\t' + str(t1_cov[ratio]['xcov']))
        to_print_too = (t1_zero_cov, t1_pct_zero_cov, t2_avg_cov, t2_avg_ratio)
        tbl.write('\t' + '\t'.join(str(e) for e in to_print_too))
        for ratio in ranges:
            tbl.write('\t' + str(t2_cov[ratio]['xcov']))
        tbl.write('\t' + '\t'.join((str(t2_zero_cov), str(t2_pct_zero_cov))) + '\n')
        tbl.close()
        json_dict = {'BionimbusID': RG[0], 'Date': RG[1], 'Machine': RG[2], 'Run': RG[3], 'BarCode': RG[4],
                     'Lane': RG[5], 'read_length': rd_len, 'total_reads': tot_rds,
                     'post_align_reads': post_aligned_reads, 'percent_aligned': frac_aligned,
                     'post_rmdup_reads': rmduped_reads, 'fraction_rmduped': frac_rmduped_reads,
                     'target_size': target_size, 'aligned_bp_ot': aligned_bp_ot,
                     'fraction_aligned_bp_ot': fraction_aligned_bp_ot,
                     'fraction_sequenced_bp_ot': fraction_sequenced_ot, 'median_insert_size': median_insert_size,
                     'median_absolute_deviation': median_absolute_deviation, 'mean_insert_size': mean_insert_size,
                     'insert_standard_deviation': insert_standard_deviation, 'date_aligned': date_aligned, 'coverage': {
                (ccvsn + '_capture_t1'): {'average': t1_avg_cov, '%covered_at_average': t1_avg_ratio,
                                          '90%': t1_cov[0.9]['xcov'], '50%': t1_cov[0.5]['xcov'],
                                          '10%': t1_cov[0.1]['xcov'], 'zero_cov_bp': t1_zero_cov,
                                          '%zero_cov': t1_pct_zero_cov},
                (ccvsn + '_capture_t2'): {'average': t2_avg_cov, '%covered_at_average': t2_avg_ratio,
                                          '90%': t2_cov[0.9]['xcov'], '50%': t2_cov[0.5]['xcov'],
                                          '10%': t2_cov[0.1]['xcov'], 'zero_cov_bp': t2_zero_cov,
                                          '%zero_cov': t2_pct_zero_cov}}}
        json_out.write(json.dumps(json_dict, indent=4, sort_keys=True))
        json_out.close()
    else:
        log(loc, date_time() + 'Whole genome flag given\n')
        bed_hist = sample + '.genome.hist'
        # cov is a dictionary with the per -ratio stats
        tbl.write(
            "BionimbusID\tDate\tMachine\tRun\tBarCode\tLane\tread_length\ttotal_reads\tpost_align_reads\t" +
            "percent_aligned\tpost_rmdup_reads\tfraction_rmduped\ttarget_size\taligned_bp_ot\tfraction_aligned_bp_ot" +
            "\tfraction_sequenced_bp_ot\tmedian_insert_size\tmedian_absolute_deviation\tmean_insert_size\t" +
            "insert_standard_devation\tdate_aligned\taverage_x_coverage\t%covered_at_average\t90%\t50%\t10%\t" +
            "zero_cov_bp\t%zero_cov\n")
        tbl.write('\t'.join(RG) + '\t' + str(rd_len))
        log(loc, 'Parsing bedtools coverage file ' + bed_hist + '\n')
        (target_size, aligned_bp_ot, num_bp, cov, zero_cov, pct_zero_cov, avg_cov, avg_ratio) = parseCoverage(bed_hist,
                                                                                                              ranges)
        fraction_aligned_bp_ot = float(num_bp) / float(target_size)
        fraction_sequenced_bp_ot = float(aligned_bp_ot) / float(all_bp_est)
        to_print = (
            tot_rds, post_aligned_reads, frac_aligned, rmduped_reads, frac_rmduped_reads, target_size, aligned_bp_ot,
            fraction_aligned_bp_ot, fraction_sequenced_bp_ot, median_insert_size, median_absolute_deviation,
            mean_insert_size,
            insert_standard_deviation, date_aligned, avg_cov, avg_ratio)
        tbl.write('\t' + '\t'.join(str(e) for e in to_print))
        for ratio in ranges:
            tbl.write('\t' + str(cov[ratio]['xcov']))
        tbl.write('\t' + "\t".join((str(zero_cov), str(pct_zero_cov))) + '\n')
        tbl.close()
        json_dict = {'BionimbusID': RG[0], 'Date': RG[1], 'Machine': RG[2], 'Run': RG[3], 'BarCode': RG[4],
                     'Lane': RG[5], 'read_length': rd_len, 'total_reads': tot_rds,
                     'post_align_reads': post_aligned_reads, 'percent_aligned': frac_aligned,
                     'post_rmdup_reads': rmduped_reads, 'fraction_rmduped': frac_rmduped_reads,
                     'target_size': target_size, 'aligned_bp_ot': aligned_bp_ot,
                     'fraction_aligned_bp_ot': fraction_aligned_bp_ot,
                     'fraction_sequenced_bp_ot': fraction_sequenced_bp_ot, 'median_insert_size': median_insert_size,
                     'median_absolute_deviation': median_absolute_deviation, 'mean_insert_size': mean_insert_size,
                     'insert_standard_deviation': insert_standard_deviation, 'date_aligned': date_aligned, 'coverage': {
                ccvsn: {'average': avg_cov, 'covered_at_average': avg_ratio, '90%': cov[0.9]['xcov'],
                        '50%': cov[0.5]['xcov'], '10%': cov[0.1]['xcov'], 'zero_cov_bp': zero_cov,
                        '%zero_cov': pct_zero_cov}}}
        json_out.write(json.dumps(json_dict, indent=4, sort_keys=True))
        json_out.close()


if __name__ == "__main__":
    import argparse

    parser = argparse.ArgumentParser(description='Coverage and algnment QC summary.')
    parser.add_argument('-j', '--json', action='store', dest='config_file',
                        help='JSON config file with tool and reference locations')
    parser.add_argument('-f', '--flag', action='store', dest='cflag',
                        help='\'y\' if whole genome, \'n\' if custom capture to provide per-tier coverage stats')
    parser.add_argument('-sa', '--sample', action='store', dest='sample', help='Sample prefix')

    if len(sys.argv) == 1:
        parser.print_help()
        sys.exit(1)

    inputs = parser.parse_args()
    (config_file, sample, cflag) = (inputs.config_file, inputs.sample, inputs.cflag)
    parse_qc(config_file, sample, cflag)
